shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -918,56 +928,65 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
             val rs = con(tym, res, res)
-            (n, FieldType(Some(rs), rs)(prov))
-          } else (n, tym.toUpper(prov))
+            (n, FieldType(Some(rs), rs)(fprov))
+          } else (n, tym.toUpper(fprov))
         })(prov)
       case tup: Tup if funkyTuples =>
         typeTerms(tup :: Nil, false, Nil)
       case Tup(fs) =>
-        TupleType(fs.map{ case (n, (t, mut)) =>
+        TupleType(fs.map { case (n, (t, mut)) =>
           val tym = typeTerm(t)
+          val fprov = tp(t.toLoc, "tuple field")
           if (mut) {
-            val res = freshVar(prov)
+            val res = freshVar(tym.prov)
             val rs = con(tym, res, res)
-            (n, FieldType(Some(rs), rs)(prov))
-          } else (n, tym.toUpper)
+            (n, FieldType(Some(rs), rs)(fprov))
+          } else (n, tym.toUpper(fprov))
         })(fs match {
           case Nil | ((N, _) :: Nil) => noProv
           case _ => tp(term.toLoc, "tuple literal")
         })
       case Subs(a, i) =>
         val t_a = typeTerm(a)
         val t_i = typeTerm(i)
-        val res = freshVar(prov)
-        con(t_a, ArrayType(res.toUpper)(prov), t_a)
-        con(t_i, IntType, t_i)
-        res
-      case Assign(s @ Sel(r, f), rhs) => 
+        con(t_i, IntType, TopType)
+        val elemType = freshVar(prov)
+        con(t_a, ArrayType(elemType.toUpper(tp(i.toLoc, "array element")))(prov), elemType)
+      case Assign(s @ Sel(r, f), rhs) =>
         val o_ty = typeTerm(r)
-        val res = freshVar(prov)
+        val sprov = tp(s.toLoc, "assigned selection")
+        val fieldType = freshVar(sprov)
+        val obj_ty =
+          // Note: this proxy does not seem to make any difference:
+          mkProxy(o_ty, tp(r.toCoveringLoc, "receiver"))
+        con(obj_ty, RecordType.mk((f, FieldType(Some(fieldType), fieldType)(
+          tp(f.toLoc, "assigned field")
+        )) :: Nil)(sprov), fieldType)
         val vl = typeTerm(rhs)
-        val obj_ty = mkProxy(o_ty, tp(r.toCoveringLoc, "receiver"))
-        val rf = con(obj_ty, RecordType.mk((f, FieldType(Some(res), res)(prov)) :: Nil)(prov), res)
-        con(vl, rf, vl)
-        UnitType
+        con(vl, fieldType, UnitType.withProv(prov))
       case Assign(s @ Subs(a, i), rhs) => 
         val a_ty = typeTerm(a)
-        val res = freshVar(prov)
+        val i_ty = typeTerm(i)
+        con(i_ty, IntType, TopType)
+        val sprov = tp(s.toLoc, "assigned array element")
+        val elemType = freshVar(sprov)
         val vl = typeTerm(rhs)
-        val arr_ty = mkProxy(a_ty, tp(a.toCoveringLoc, "receiver"))
-        con(arr_ty, ArrayType(FieldType(Some(res), res)(prov))(prov), res)
-        con(vl, IntType, vl)
-        UnitType
-      case Assign(_, rhs) => 
-        err(msg"Cannot not assign to non-field", term.toLoc)(raise)
+        val arr_ty =
+            // Note: this proxy does not seem to make any difference:
+            mkProxy(a_ty, tp(a.toCoveringLoc, "receiver"))
```

Comment:
I actually don't know what this is proxy is doing, as I just copied that from `rcdSel` @_@, since the typing looks similar